### PR TITLE
add unit tests for pkg/api/resource.go

### PR DESCRIPTION
**Pull request by concaf**
**What this PR does / why we need it**:
This PR adds tests for pkg/api/resource.go in pkg/api/resource_test.go.

**Special notes for your reviewer**:
FWIW this is my first PR to Kubernetes, so apologies if I messed up somewhere or didn't follow some guidelines :)

### CODE REVIEW

**Comment by xiangpengzhao** [comment]
/release-note-none
/ok-to-test

Awesome! Thanks for contribution!

**Comment by gaocegege** [comment]
Maybe you should run `./hack/update-bazel.sh` to pass `pull-kubernetes-verify`

**Comment by containscafeine** [comment]
@gaocegege thank you, that fixed the problem :)

**Comment by containscafeine** [comment]
Hey @erictune @luxas, could you PTAL? Is there something I need to change in this? TIA :)

**Comment by luxas** [comment]
Reassigning to @sttts from my side.
/unassign
/assign @sttts 

FWIW; I didn't see anything wrong with this PR at a first glance ;)

File: pkg/api/resource_test.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+func TestResourceNameString(t *testing.T) {
+	testCases := []struct {
+		resourceName       ResourceName
+		resourceNameString string
+	}{
+		{
+			resourceName:       ResourceName("secret"),
```
Comment (sttts) [inline pkg/api/resource_test.go]: please split the input of the test and the actual test procedure, i.e. move the func call down into the loop.

File: pkg/api/resource_test.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+func TestResourceListCpu(t *testing.T) {
+	testCases := []struct {
+		description      string
+		resourceList     ResourceList
+		resourceQuantity resource.Quantity
+	}{
+		{
+			description:      "ResourceCPU is present in ResourceList",
+			resourceList:     ResourceList{ResourceCPU: resource.MustParse("100m")},
+			resourceQuantity: resource.MustParse("100m"),
```
Comment (sttts) [inline pkg/api/resource_test.go]: please call the expected outcome `expected`. It would make it much easier to understand what the test is actually doing.

**Comment by erictune** [comment]
/approve
will let sttts give the lgtm

**Comment by sttts** [comment]
pkg/api moved to pkg/apis/core.

File: pkg/api/resource_test.go (hunk 2 above)
Comment (containscafeine) [reply pkg/api/resource_test.go]: done, changed in all the test structs in the file

File: pkg/api/resource_test.go (hunk 1 above)
Comment (containscafeine) [reply pkg/api/resource_test.go]: I didn't understand this part :(

In the part -
```golang
	testCases := []struct {
		resourceName ResourceName
		expected     string
	}{
		{
			resourceName: ResourceName("secret"),
			expected:     "secret",
		},
	}

	for _, tc := range testCases {
		if tc.resourceName.String() != tc.expected {
			t.Errorf("expected ResourceName string to be '%v', but got '%v' instead", tc.expected, tc.resourceName.String())
		}
	}
```

Here, the test procedure, i.e.
```golang
if tc.resourceName.String() != tc.expected {
```
is in the loop while the test input -
```golang
	testCases := []struct {
		resourceName ResourceName
		expected     string
	}{...}
```
is out of the loop.

Could you explain it a bit more?

**Review by concaf — COMMENTED** [review]
@sttts, rebased and made the refactor. Didn't understand one of your comments https://github.com/kubernetes/kubernetes/pull/52716#discussion_r148454605, though, commented there. Thanks for the review :tada:

File: pkg/apis/core/resource_test.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	}{
+		{
+			resourceName: ResourceName("secret"),
+			expected:     "secret",
+		},
+	}
+
+	for _, tc := range testCases {
+		if tc.resourceName.String() != tc.expected {
+			t.Errorf("expected ResourceName string to be '%v', but got '%v' instead", tc.expected, tc.resourceName.String())
```
Comment (sttts) [inline pkg/apis/core/resource_test.go]: Instead of `'%v'` better write `%q`. It does escaping correctly if necessary.

Comment (sttts) [reply pkg/apis/core/resource_test.go]: nit: as there is only one test, I would also not mind to drop the loop.

File: pkg/apis/core/resource_test.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	}{
+		{
+			description:  "ResourceCPU is present in ResourceList",
+			resourceList: ResourceList{ResourceCPU: resource.MustParse("100m")},
+			expected:     resource.MustParse("100m"),
+		},
+		{
+			description:  "ResourceCPU is not present in ResourceList",
+			resourceList: ResourceList{ResourceConfigMaps: resource.MustParse("10")},
+			expected:     resource.Quantity{Format: resource.DecimalSI},
```
Comment (sttts) [inline pkg/apis/core/resource_test.go]: the `10` is dropped?

Comment (sttts) [reply pkg/apis/core/resource_test.go]: oh ic, you ask for CPU

File: pkg/api/resource_test.go (hunk 1 above)
Comment (sttts) [reply pkg/api/resource_test.go]: Looks good. I was assuming that you test `ResourceName`, but you test `String`.

File: pkg/apis/core/resource_test.go (hunk 3 above)
Comment (sttts) [reply pkg/apis/core/resource_test.go]: There are some more `'%v'` below.

**Comment by sttts** [comment]
Some nits. @erictune feel free to tag, when the nits are solved.

File: pkg/apis/core/resource_test.go (hunk 3 above)
Comment (containscafeine) [reply pkg/apis/core/resource_test.go]: Ack.
- replaced `'%v`s with `%q`
- dropped the loop

**Review by concaf — COMMENTED** [review]
@sttts @erictune made the changes, PTAL :)

File: pkg/apis/core/resource_test.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+import (
+	"reflect"
+	"testing"
+
+	"github.com/davecgh/go-spew/spew"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+func TestResourceNameString(t *testing.T) {
+	testCase := struct {
```
Comment (sttts) [inline pkg/apis/core/resource_test.go]: no need for the testcase struct ;-)

File: pkg/api/resource_test.go (hunk 1 above)
Comment (containscafeine) [reply pkg/api/resource_test.go]: ack

File: pkg/apis/core/resource_test.go (hunk 5 above)
Comment (containscafeine) [reply pkg/apis/core/resource_test.go]: ack, removed

Comment (containscafeine) [reply pkg/apis/core/resource_test.go]: ack, removed

**Comment by containscafeine** [comment]
@sttts made the changes, PTAL! Sorry for the late reply.

**Comment by containscafeine** [comment]
ping @sttts, gentle reminder

**Pull request closed without merging**